nixos/lib/test-driver/test_driver/qmp.py

```diff
@@ -0,0 +1,65 @@
+import socket
+from queue import Queue
+from typing import Dict, Any, Iterator
+import json
+from pathlib import Path
+
+
+class QMPSession:
+    def __init__(self, sock: socket.socket) -> None:
+        self.sock = sock
+        self.pending_events: Queue[Dict[str, Any]] = Queue()
+        self.reader = sock.makefile("r")
+        self.writer = sock.makefile("w")
+        hello = self._result()
+        assert "QMP" in hello, f"Unexpected result: {hello}"
+        self.send("qmp_capabilities")
+
+    @classmethod
+    def from_path(cls, path: Path) -> "QMPSession":
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(str(path))
+        return cls(sock)
+
+    def __del__(self) -> None:
+        self.sock.close()
+
+    def _readmsg(self) -> Dict[str, Any]:
+        line = self.reader.readline()
+        return json.loads(line)
+
+    def _raise_unexpected_msg(self, msg: Dict[str, Any]) -> None:
+        m = json.dumps(msg, sort_keys=True, indent=4)
+        raise RuntimeError(f"Got unexpected qmp response: {m}")
```

Comment:
I don't like that this class serializes useful information into a string and then throws it as a generic runtime error.

Can this please create a completely new class that:

- inherits from runtime error
- provides a nice [`__str__`](https://docs.python.org/3/reference/datamodel.html#object.__str__) method for everyone who just wants to print it
- but also provides access to the actual object as an object

This way users can decide themselves if they want to just print stuff or actually look into the error programmatically, instead of this chance be taken away due to premature serialization.

Reply:
I tried something, let me know how you like it.